emulator/src/uartlite.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include "mmio_dev.h"
+#include <cstring>
+#include <algorithm>
+#include <queue>
+#include <mutex>
+
+#define SR_TX_FIFO_FULL         (1<<3) /* transmit FIFO full */
+#define SR_TX_FIFO_EMPTY        (1<<2) /* transmit FIFO empty */
+#define SR_RX_FIFO_VALID_DATA   (1<<0) /* data in receive FIFO */
+#define SR_RX_FIFO_FULL         (1<<1) /* receive FIFO full */
+
+#define ULITE_CONTROL_RST_TX	0x01
+#define ULITE_CONTROL_RST_RX	0x02
+
+struct uartlite_regs {
+    unsigned int rx_fifo;
+    unsigned int tx_fifo;
+    unsigned int status;
+    unsigned int control;
+};
+
+class uartlite : public mmio_dev  {
+public:
+    uartlite() {
+        memset(&regs,0,sizeof(regs));
+        regs.status = SR_TX_FIFO_EMPTY;
+        wait_ack = false;
+    }
+    bool do_read(uint64_t start_addr, uint64_t size, unsigned char* buffer) {
+        std::unique_lock<std::mutex> lock(rx_lock);
+        if (start_addr + size > sizeof(regs)) return false;
+        if (!rx.empty()) {
+            regs.status |= SR_RX_FIFO_VALID_DATA;
+            regs.rx_fifo = rx.front();
+        }
+        else regs.status &= ~SR_RX_FIFO_VALID_DATA;
+        memcpy(buffer,((char*)(&regs))+start_addr,std::min(size,sizeof(regs)-start_addr));
```

Comment:
`size` is already checked at line 33, the `std::min` here seems redundant, we can use size directly. Same on `do_write`.